### PR TITLE
Megatron style TFLOPs Calculation

### PR DESCRIPTION
**Pull request by abhinavgoel95**
@rwitten this is a draft.

This type of change would be specific to a few transformer models (e.g., Gemma, LLama, GPT, etc.). It wouldn't work with MoE, or some new architectures.

I was thinking that walking through the train-step and calculating the FLOPs layer-by-layer would be a very intrusive change.

What do you think?

### CODE REVIEW

File: MaxText/maxtext_utils.py [hunk 1]
```diff
@@ -92,11 +92,24 @@ def get_train_input_output_trees(func, input_args, input_kwargs):
 # https://arxiv.org/pdf/2204.02311.pdf Appendix B
 def calculate_tflops_training_per_device(num_model_parameters, config, log=True):
   """ Calculate training TFLOP"""
-  learnable_weight_tflops = 6 * num_model_parameters * config.max_target_length * config.per_device_batch_size \
+  if config.use_megatron_tflops:
+    attention_flops = 2 * config.per_device_batch_size * config.max_target_length**2 * config.num_query_heads * config.head_dim
+    ffn1_flops = 2 * config.per_device_batch_size * config.max_target_length * config.mlp_dim * config.emb_dim * len(config.mlp_activations)
+    ffn2_flops = 2 * config.per_device_batch_size * config.max_target_length * config.mlp_dim * config.emb_dim
+    qkv_flops = 6 * config.per_device_batch_size * config.max_target_length * config.emb_dim * config.num_query_heads * config.head_dim
```
Comment (rwitten) [inline MaxText/maxtext_utils.py:99]: Watch out! This should be
```
2 * config.per_device_batch_size * config.max_target_length * config.emb_dim * (config.num_query_heads + 2 * config.num_kv_heads) * config.head_dim
```

I don't think this resolves our gap because llama7b is MHA

File: MaxText/maxtext_utils.py [hunk 2]
```diff
@@ -92,11 +92,24 @@ def get_train_input_output_trees(func, input_args, input_kwargs):
 # https://arxiv.org/pdf/2204.02311.pdf Appendix B
 def calculate_tflops_training_per_device(num_model_parameters, config, log=True):
   """ Calculate training TFLOP"""
-  learnable_weight_tflops = 6 * num_model_parameters * config.max_target_length * config.per_device_batch_size \
+  if config.use_megatron_tflops:
+    attention_flops = 2 * config.per_device_batch_size * config.max_target_length**2 * config.num_query_heads * config.head_dim
```
Comment (rwitten) [inline MaxText/maxtext_utils.py:96]: Am I right in understanding that Megatron is counting attention flops non-causally and MaxText is counting it causally?

Comment (rwitten) [reply MaxText/maxtext_utils.py:96]: I really want to cut down on confusion but ... isn't attention causal?

File: MaxText/maxtext_utils.py [hunk 3]
```diff
@@ -92,11 +92,24 @@ def get_train_input_output_trees(func, input_args, input_kwargs):
...
+    ffn1_flops = 2 * config.per_device_batch_size * config.max_target_length * config.mlp_dim * config.emb_dim * len(config.mlp_activations)
+    ffn2_flops = 2 * config.per_device_batch_size * config.max_target_length * config.mlp_dim * config.emb_dim
+    qkv_flops = 6 * config.per_device_batch_size * config.max_target_length * config.emb_dim * config.num_query_heads * config.head_dim
+    proj_flops = 2 * config.per_device_batch_size * config.max_target_length * config.emb_dim * config.num_query_heads * config.head_dim
+
+    # multiplied by 3 to account for fprop and bprop flops
+    learnable_weight_tflops = (ffn1_flops + ffn2_flops + qkv_flops + proj_flops) * config.num_decoder_layers * 3 / 10**12
+    causal_attention_tflops = attention_flops * config.num_decoder_layer * 3 / 10**12
```
Comment (rwitten) [inline MaxText/maxtext_utils.py:104]: How does Megatron account for the unprojection matmul? I.e. last layer output to logits.

**Review by rwitten — CHANGES_REQUESTED** [review]

**Comment by abhinavgoel95** [comment]
Made the changes as requested in the meeting @rwitten

File: MaxText/maxtext_utils.py (hunk 2 above)
Comment (abhinavgoel95) [reply MaxText/maxtext_utils.py]: Decided in the meeting that we would ignore causality to match Megatron.

**Comment by abhinavgoel95** [comment]
cc @rwitten following up on this
